lib/upnp/upnp.go

```diff
@@ -96,19 +98,32 @@ func Discover(ctx context.Context, _, timeout time.Duration) []nat.Device {
 	resultChan := make(chan nat.Device)
 
 	wg := &sync.WaitGroup{}
+	wgforIPv4 := &sync.WaitGroup{}
 
 	for _, intf := range interfaces {
 		if intf.Flags&net.FlagRunning == 0 || intf.Flags&net.FlagMulticast == 0 {
 			continue
 		}
 
+		wg.Add(3)
+		wgforIPv4.Add(2)
 		for _, deviceType := range []string{"urn:schemas-upnp-org:device:InternetGatewayDevice:1", "urn:schemas-upnp-org:device:InternetGatewayDevice:2"} {
-			wg.Add(1)
 			go func(intf net.Interface, deviceType string) {
-				discover(ctx, &intf, deviceType, timeout, resultChan)
+				// For each protocol, try to discover IPv6 gateways.
+				discover(ctx, &intf, deviceType, timeout, resultChan, false)
 				wg.Done()
+				wgforIPv4.Done()
 			}(intf, deviceType)
+
 		}
+
+		wgforIPv4.Wait()
```

Comment:
Hm, you're right, but also the code is almost completely sequential now (the only thing happening in parallel is IPv4 IGDv1 and IGDv2). Since this is all likely to happen over the local network and already uses timeouts, the delay should be negligible. Should we just make it sequential and don't use goroutines or synchronization at all?

Reply:
Or maybe we should use one goroutine per interface and have it go through all three IGD versions.

Reply:
> Or maybe we should use one goroutine per interface and have it go through all three IGD versions.

I've implemented that approach now, let me know if you think this might still be problematic

Reply:
One goroutine per interface sounds good. We end up with lower complexity, dodge buggy devices and it should still be fast.